sgx-sys/src/lib.rs

```diff
@@ -12,80 +12,104 @@
-//! https://www.intel.com/content/dam/www/public/emea/xe/en/documents/manuals/64-ia-32-architectures-software-developer-vol-3d-part-4-manual.pdf
-
 #![deny(clippy::all)]
 #![allow(clippy::identity_op)]
 
+mod ioctl;
+
 use ioctl::Ioctl;
 use std::fs::File;
 use std::io::Result;
-use std::mem::size_of;
+use std::mem::size_of_val;
+use std::marker::PhantomData;
 
-mod ioctl;
-pub mod secs;
-pub mod sigstruct;
-pub mod tcs;
-
-pub struct EnclaveBuilder(File);
-pub struct Enclave(File);
-
-pub enum Page<'a> {
-    Tcs(&'a tcs::Tcs),
-    Regular {
-        data: &'a [u8],
-        perms: ioctl::sgx::PagePerms,
-        flags: ioctl::sgx::PageFlags,
-    },
-}
+use sgx_traits::{Handle, Flags, Builder as BuilderTrait, Enclave as EnclaveTrait};
+use sgx_types::secinfo::{SecInfo, Flags as Perms, PageType};
+use sgx_types::sigstruct::SigStruct;
+use sgx_types::secs::Secs;
+use sgx_types::tcs::Tcs;
+use paged::{Page, Size4k};
+
+pub struct Builder<'b>(File, PhantomData<&'b ()>);
+
+impl<'b> BuilderTrait<'b> for Builder<'b> {
+    type Enclave = Enclave<'b>;
 
-impl EnclaveBuilder {
     // Calls SGX_IOC_ENCLAVE_CREATE (ECREATE)
-    pub fn new(secs: &secs::Secs) -> Result<EnclaveBuilder> {
-        let create = ioctl::sgx::Create::new(secs);
+    fn new(secs: Secs) -> Result<Self> {
+        let page = secs.into();
+        let create = ioctl::sgx::Create::new(&page);
         let file = File::open("/dev/sgx/enclave")?;
         create.ioctl(&file)?;
-        Ok(Self(file))
+        Ok(Self(file, PhantomData))
     }
 
-    // Calls SGX_IOC_ENCLAVE_ADD_PAGES (EADD)
-    pub fn add_pages(self, page: Page, offset: usize) -> Result<Self> {
-        use ioctl::sgx;
-
-        let (data, si, flags) = match &page {
-            Page::Tcs(tcs) => {
-                let data = unsafe {
-                    std::slice::from_raw_parts(
-                        *tcs as *const tcs::Tcs as *const u8,
-                        size_of::<tcs::Tcs>(),
-                    )
-                };
-
-                let si = sgx::SecInfo::new(
-                    sgx::PagePerms::READ | sgx::PagePerms::WRITE,
-                    sgx::PageType::Tcs,
-                );
-
-                (data, si, sgx::PageFlags::MEASURE)
-            }
-
-            Page::Regular { data, perms, flags } => {
-                (*data, sgx::SecInfo::new(*perms, sgx::PageType::Reg), *flags)
-            }
+    fn add_tcs(
+        &mut self,
+        tcs: Tcs,
+        offset: usize
+    ) -> Result<Handle<'b, Tcs>> {
+        let page: Page<Size4k, _> = tcs.into();
+
+        let data = unsafe {
+            std::slice::from_raw_parts(
+                page.as_ref() as *const Tcs as *const u8,
+                size_of_val(&page),
+            )
+        };
+
+        let si = SecInfo::new(Perms::R | Perms::W, PageType::Tcs);
+        self.add_slice(&data, offset, Flags::MEASURE, si)
+            .map(|_| unsafe { Handle::new(offset) })
+    }
+
+    fn add_struct<T>(
```

Comment:
I would like to understand a bit more about how this function `add_struct` differs from `add_slice` down below and the use cases for each. Why are they two separate functions? And why are they both different from `add_tcs`? I don't yet understand the different behavior we want to get from each (although I see the use of the new `Handle` and `Page<Size4k, _>`).

Reply:
Basically, everything is just `add_slice()`. The `add_struct()` and `add_tcs()` functions are just convenience wrappers.